rust/tools/authorization-logic/src/souffle/lowering_ast_datalog.rs

```diff
@@ -102,6 +102,10 @@
 use crate::{ast::*, souffle::datalog_ir::*};
 use std::collections::HashMap;
 
+fn pred_to_dlir_rvalue(pred: &AstPredicate) -> DLIRRvalue {
+    DLIRRvalue::PredicateRvalue { predicate: pred.clone() }
+}
```

Comment:
You could make this method require an `AstPredicate` rather than an `&AstPredicate`. That would allow you to skip cloning at call sites where the `pred` isn't needed for anything anymore anyway, which could save some copies. Also, might make the borrow checker easier to placate at callsites.

Reply:
This gets called inside `map` combinators, and in rust the arguments to these are `&` so that makes things more awkward and doesn't allow the clone to be skipped (it just requires cloning in the map).